ARROW-6229: [C++] add a directory based data source implementation

bkietz opened the pull request:
Adds DirectoryBasedDataSource, which is constructed with a file format and reads a given directory recursively on construction then yields data fragments from discovered files matching that format. 

Also adds FileFormat::MakeFragment which creates a(n instance of a subclass of) FileBasedDataFragment given a FileSource and ScanOptions.

fsaintjacques commented on cpp/src/arrow/dataset/file_base.h:
```diff
@@ -129,25 +130,59 @@ class ARROW_DS_EXPORT FileFormat {
-  std::shared_ptr<FileFormat> format() const { return format_; }
+/// \brief A DataSource which takes files of one format from a directory
+///
+/// The directory is read on construction and files are not checked for existence
```
> The directory is crawled upon construction (Make) and not updated afterward, e.g. it will not take new files and will error if files are deleted/moved.

bkietz replied:
Thanks

fsaintjacques commented on cpp/src/arrow/filesystem/path-util.h:
```diff
@@ -36,6 +36,10 @@ constexpr char kSep = '/';
 
+// Return the extension of the file
+ARROW_EXPORT
+std::string GetAbstractPathExtension(const std::string& s);
```
Add a small unit test for this.

bkietz replied:
Will do

fsaintjacques commented on cpp/src/arrow/util/iterator.h:
```diff
@@ -32,7 +32,7 @@ template <typename T>
  public:
   static_assert(std::is_assignable<T, std::nullptr_t>::value,
-                "NULL is used to signal completion");
+                "nullptr is used to signal completion");
```
Linter complained about this, hence NULL.

bkietz replied:
Aw, that's brain dead. I'll revert it

fsaintjacques replied:
yeah, linter could be more intelligent, but heh.

fsaintjacques commented on cpp/src/arrow/dataset/file_base.cc:
```diff
@@ -41,5 +44,54 @@ Status FileBasedDataFragment::GetTasks(std::shared_ptr<ScanContext> scan_context
+  std::vector<fs::FileStats> stats;
+  fs::Selector s;
+  s.base_dir = location.path();
+  s.recursive = true;
```
This option needs to be exposed to the `DirectoryBasedDataSource`. Make a struct with options, it'll serve for other stuff too.

fsaintjacques commented on cpp/src/arrow/dataset/file_base.h:
```diff
@@ -129,25 +130,59 @@ class ARROW_DS_EXPORT FileFormat {
+///
+/// The directory is read on construction and files are not checked for existence
+/// thereafter.
+class ARROW_DS_EXPORT DirectoryBasedDataSource : public DataSource {
```
`FileSystemDataSource` might be a better name?

fsaintjacques commented on cpp/src/arrow/dataset/file_base.h:
```diff
@@ -129,25 +130,59 @@ class ARROW_DS_EXPORT FileFormat {
   FileSource location_;
+  std::vector<fs::FileStats> stats_;
   std::shared_ptr<FileFormat> format_;
   std::shared_ptr<ScanOptions> scan_options_;
```
Is this used, can we drop it for now?

bkietz replied:
This is used to store the paths of files found on construction

fsaintjacques replied:
The `scan_options`?

bkietz replied:
ah, sorry
*that* is passed to FileFormat::MakeFragment in GetFragments

fsaintjacques commented on cpp/src/arrow/dataset/file_parquet-test.cc:
```diff
@@ -81,5 +82,26 @@ TEST_F(TestParquetFileFormat, ScanRecordBatchReader) {
+  TestDirectorBasedDataSource() {}
+};
+
+TEST_F(TestDirectorBasedDataSource, Directory) {
```
I'd want more test for this, notably test with and without recursion (and more than 1 level).

fsaintjacques commented on cpp/src/arrow/dataset/file_base.cc:
```diff
@@ -41,5 +44,54 @@ Status FileBasedDataFragment::GetTasks(std::shared_ptr<ScanContext> scan_context
+    std::unique_ptr<FileBasedDataFragment> fragment;
+    auto st = format_->MakeFragment(l, scan_options_, &fragment);
+    if (!st.ok()) {
+      return internal::make_unique<EmptyIterator<std::shared_ptr<DataFragment>>>(st);
```
Hum, I'm not sure I agree. Our choices are:

- Change the signature to return failure (possible)
- Construct the DataFragments at construction, the downside is that a data fragment and/or data source might only be consumed once. This is a general problem we might have to take a decision.

The solution to returns an EmptyIterator seems like a hack. I would think the first option is probably better, because the data source _can_ fail due to external reasons, e.g. file system activity.

bkietz replied:
I don't think the second option is viable; if DataSources are single use then that would apply to Datasets as well. Would it be more acceptable to lazily make fragments in the iterator (rather than making them all ahead of time)? I can also change the return type to `Result<ptr<DataFragmentIterator>>` if necessary, but it seems reasonable to defer error handling to the data fragment consumer.

bkietz replied:
I don't think the second option is viable; if DataSources are single use then that would apply to Datasets as well. Would it be more acceptable to lazily make fragments in the iterator (rather than making them all ahead of time)? I can also change the return type to Result<ptr<DataFragmentIterator>> if necessary, but it seems reasonable to defer error handling to the data fragment consumer.

fsaintjacques replied:
I think we should either use all status or all result, but not a mix.

fsaintjacques replied:
Also note that while it must yield an iterator, the failure could happen in the first call of Next.

fsaintjacques replied:
I think I prefer this version:

- Combine VectorIterator(paths) with MapIterator(pathToFragment, paths_it)

bkietz replied:
done

fsaintjacques commented on cpp/src/arrow/dataset/file_base.cc:
```diff
@@ -41,5 +44,54 @@ Status FileBasedDataFragment::GetTasks(std::shared_ptr<ScanContext> scan_context
+  auto new_end =
+      std::remove_if(stats.begin(), stats.end(), [&](const fs::FileStats& stats) {
+        return stats.type() != fs::FileType::File ||
+               !format->IsKnownExtension(stats.extension());
```
IMHO, globbing is much better than extension based discovery. I'd say open a high priority followup ticket or implement it immediately.

bkietz replied:
[picomatch](https://github.com/micromatch/picomatch) is MIT licensed, I could port it and add it to the filesystem API

bkietz replied:
@pitrou could you comment?

bkietz replied:
https://issues.apache.org/jira/browse/ARROW-6257

bkietz commented:
@fsaintjacques how's that?

fsaintjacques commented on cpp/src/arrow/dataset/file_base.cc:
```diff
@@ -41,5 +44,56 @@ Status FileBasedDataFragment::Scan(std::shared_ptr<ScanContext> scan_context,
+  auto scan_options = scan_options_;
+  auto format = format_;
+  auto filesystem = filesystem_;
+  return MakeIterator([=](std::shared_ptr<DataFragment>* out) mutable {
```
The complexity of such succinct code is very high (4 captures, lifetime requirement, mutable!, output parameter...). I think it warrants creating an explicit Iterator class.

bkietz replied:
Okay, can do.

fsaintjacques commented on cpp/src/arrow/dataset/file_base.h:
```diff
@@ -147,11 +153,43 @@ class ARROW_DS_EXPORT FileBasedDataFragment : public DataFragment {
   const FileSource& source() const { return source_; }
   std::shared_ptr<FileFormat> format() const { return format_; }
 
+  std::shared_ptr<ScanOptions> scan_options() const override { return scan_options_; }
```
Shouldn't file format be stateless?

bkietz replied:
If so, then we could replace `std::shared_ptr<FileFormat>` with `const FileFormat*` everywhere and let all FileFormat implementations be singletons. Is that what you're suggesting?

fsaintjacques commented on cpp/src/parquet/arrow/dataset/file_parquet.cc:
```diff
@@ -0,0 +1,165 @@
+namespace arrow {
+namespace dataset {
+
+using ScanTaskPtr = std::unique_ptr<::arrow::dataset::ScanTask>;
```
Seems like you rebased it to the wrong file path :)

bkietz replied:
darn. I'll fix it

wesm committed:
ARROW-5480: [Python] Add unit test asserting specifically that pandas.Categorical roundtrips to Parquet format without special options

This only works for string types for the moment. Once ARROW-6277 is addressed we can expand to other types.

Closes #5110 from wesm/ARROW-5480 and squashes the following commits:

a161b2432 <Wes McKinney> Add missing pandas marks
f1f808235 <Wes McKinney> Don't use pandas's Parquet functions since they don't work in CI for some reason
9e984043a <Wes McKinney> Improve unit test for out-of-order values, nulls, unobserved category values
620b3b858 <Wes McKinney> Add unit test for ARROW-5480

Authored-by: Wes McKinney <wesm+git@apache.org>
Signed-off-by: Wes McKinney <wesm+git@apache.org>

wesm committed:
ARROW-3652: [Python][Parquet] Add unit test exhibiting that pandas.CategoricalIndex survives roundtrip to Parquet format

This was addressed by ARROW-3246 and associated issues, but this adds a unit test asserting the expected behavior viz-a-viz a CategoricalIndex (since pandas row indexes go through various gymnastics on their way to and from Parquet files)

Closes #5117 from wesm/ARROW-3652 and squashes the following commits:

c98461c8a <Wes McKinney> Add unit test exhibiting that CategoricalIndex survives roundtrip to Parquet format

Authored-by: Wes McKinney <wesm+git@apache.org>
Signed-off-by: Wes McKinney <wesm+git@apache.org>

pitrou committed:
PARQUET-1640: [C++] Fix crash in parquet-encoding-benchmark

The crash is due to undefined order of global variable destruction.
But destroying a PoolBuffer relies on its MemoryPool being still alive.
Release resources in TearDown() instead.

Closes #5120 from pitrou/PARQUET-1640-encoding-benchmark-crash and squashes the following commits:

a367b5ea3 <Antoine Pitrou> PARQUET-1640:  Fix crash in parquet-encoding-benchmark

Authored-by: Antoine Pitrou <antoine@python.org>
Signed-off-by: Wes McKinney <wesm+git@apache.org>

Tomme committed:
ARROW-3538: [Python] ability to override the automated assignment of uuid for filenames when writing datasets

Implementation for: [ARROW-3538](https://issues.apache.org/jira/browse/ARROW-3538).
Allowing one to overwrite datasets rather than always creating a new copy or appending.

Closes #4630 from Tomme/ARROW-3538 and squashes the following commits:

271e0e234 <Thomas Elvey> ARROW-3538:  ability to override the automated assignment of uuid for filenames when writing datasets
1cee98bd5 <Thomas Elvey> ARROW-3538:  ability to override the automated assignment of uuid for filenames when writing datasets

Authored-by: Thomas Elvey <tomelvey@googlemail.com>
Signed-off-by: Micah Kornfield <emkornfield@gmail.com>

lidavidm committed:
ARROW-5978: [FlightRPC] [Java] Properly release buffers in Flight integration client

Fixes a bug where dictionaries weren't properly released in cleaning up a flight stream.

Travis build: https://travis-ci.com/lihalite/arrow/builds/119807464

Recreated from https://github.com/apache/arrow/pull/4905

Closes #4913 from lihalite/flight-leak and squashes the following commits:

ac8ba8df1 <David Li> Improve documentation/tests for FlightStream dictionary provider
bca02a71f <David Li> Add test case for freeing dictionaries in Flight
a096a8097 <David Li> Properly release buffers in Flight integration client

Authored-by: David Li <li.davidm96@gmail.com>
Signed-off-by: Micah Kornfield <emkornfield@gmail.com>

wesm committed:
ARROW-6046: [C++] Do not write excess varbinary offsets in IPC messages from sliced BinaryArray

Even when creating very small slices of a large array starting from offset 0, we were writing the entire offsets buffer unnecessarily.

Closes #5126 from wesm/ARROW-6046 and squashes the following commits:

f3282f419 <Wes McKinney> Do not write excess varbinary offsets in IPC messages from sliced BinaryArray

Authored-by: Wes McKinney <wesm+git@apache.org>
Signed-off-by: François Saint-Jacques <fsaintjacques@gmail.com>

wesm committed:
ARROW-5966: [Python] Also use ChunkedStringBuilder when converting NumPy string types to Arrow StringType

We had been successfully producing chunked binary arrays but not chunked string/utf8 arrays.

Closes #5122 from wesm/ARROW-5966 and squashes the following commits:

27b062710 <Wes McKinney> Also use ChunkedStringBuilder when converting NumPy unicode types to utf8 type

Authored-by: Wes McKinney <wesm+git@apache.org>
Signed-off-by: François Saint-Jacques <fsaintjacques@gmail.com>

wesm committed:
ARROW-6048: [C++] Add ChunkedArray::View method that dispatches to Array::View

In the case where there are no chunks, it is not possible at the moment to validate whether a view would have been compatible if there had been chunks. Not sure if we want to fix that now

Closes #5127 from wesm/ARROW-6048 and squashes the following commits:

45d20f3a0 <Wes McKinney> Add ChunkedArray::View method that dispatches to Array::View

Authored-by: Wes McKinney <wesm+git@apache.org>
Signed-off-by: François Saint-Jacques <fsaintjacques@gmail.com>

fsaintjacques committed:
ARROW-6161: [C++][Dataset] Implements ParquetFragment

This implements the ParquetFragment whose main goal is to expose a streaming iterator of RecordBatch.

Closes #5083 from fsaintjacques/ARROW-6161-parquet-scan and squashes the following commits:

3fc7ce69c <Benjamin Kietzman> ignore lint for awful doxygen section
f28ef3206 <Benjamin Kietzman> hide problematic inheritance from doxygen
2d8695463 <François Saint-Jacques> Remove test_data.h.in
a9b7b91c2 <François Saint-Jacques> Follow file convention
022527f79 <Benjamin Kietzman> explicit move return for Result return
f7a28d19e <Benjamin Kietzman> clang-format
21a2be9b9 <Benjamin Kietzman> iwyu
870cfb619 <Benjamin Kietzman> move WriteRecordBatch to dataset/test_utils.h
4e480b36e <Benjamin Kietzman> replace RepeatedRecordBatch with GeneratedRecordBatch
f0ee598c0 <Benjamin Kietzman> simplify repetetive array building
82f408d3b <Benjamin Kietzman> move dataset test utilities under arrow/dataset/
b2ead08c7 <Benjamin Kietzman> include iterator
02c2c8c70 <Benjamin Kietzman> add method impls to make MSVC happy
cccb62d5f <Benjamin Kietzman> remove AllColumnIndices and AllRowGroups, add arrow::internal::Iota
c88761ff5 <Benjamin Kietzman> use RecordBatchIterator typedef
88adb7de7 <Benjamin Kietzman> first half of review comments
88f71184f <Benjamin Kietzman> include dataset.h
fa3d5cc54 <Benjamin Kietzman> move destructor to source file for MSVC
6c7d18b81 <François Saint-Jacques> Formating
c43a3a952 <François Saint-Jacques> Fix ASAN reported issue
016fa86c0 <François Saint-Jacques> Remove superfluous destructor
3c893c02c <François Saint-Jacques> Fix linter
ddbedd8b3 <François Saint-Jacques> Make gcc-5 happy
1bee33c29 <François Saint-Jacques> cmake-format
c3d840a3c <Benjamin Kietzman> refactoring, recommenting
189d11ed5 <François Saint-Jacques> Refactor FileSource::Open to switch
6e19abeea <François Saint-Jacques> Improve ScanOptions
0ce568481 <François Saint-Jacques> Fix unused variables
ff4887720 <François Saint-Jacques> Format and lint
e01dec8e2 <François Saint-Jacques> Refactor test to use in-memory RecordBatchReader
1c960be28 <François Saint-Jacques> Fix test
054bfd720 <Benjamin Kietzman> add repeatedrecordbatch
9fdded8de <François Saint-Jacques> Add ParquetFragment integration test
d951aba22 <François Saint-Jacques> Add integration data fixture class
a80503a12 <François Saint-Jacques> Move back code into arrow/dataset and add proper cmake configuration
8614626c1 <François Saint-Jacques> Refactor parquet::arrow::FileReader
aa24c1fe9 <Benjamin Kietzman> merge conflict/lint fix
d469ce042 <Benjamin Kietzman> pass Iterators by unique_ptr
fa94688dd <Benjamin Kietzman> make RecordBatchReader a refinement of the Iterator interface
5fbc38b35 <François Saint-Jacques> Initial commit

Lead-authored-by: François Saint-Jacques <fsaintjacques@gmail.com>
Co-authored-by: Benjamin Kietzman <bengilgit@gmail.com>
Signed-off-by: Wes McKinney <wesm+git@apache.org>

bkietz committed:
implement FileSystemBasedDataSource

bkietz commented:
closing in favor of https://github.com/apache/arrow/pull/5139
